walnuts_cpp/include/walnuts/nuts.hpp

```diff
@@ -98,19 +102,23 @@ void leapfrog(const F &logp_grad_fun, const Vec<S> &inv_mass, S step,
   logp_next += logp_momentum(rho_next, inv_mass);
 }
 
-template <typename S>
-bool uturn(const Span<S> &span_bk, const Span<S> &span_fw,
-           const Vec<S> &inv_mass) {
+template <Direction D, typename S>
+inline bool uturn(const Span<S> &span_1, const Span<S> &span_2,
+                  const Vec<S> &inv_mass) {
+  auto &&span_bk = (D == Direction::Forward) ? span_1 : span_2;
+  auto &&span_fw = (D == Direction::Forward) ? span_2 : span_1;
   auto scaled_diff =
       (inv_mass.array() * (span_fw.theta_fw_ - span_fw.theta_bk_).array())
           .matrix();
   return span_fw.rho_fw_.dot(scaled_diff) < 0 ||
          span_bk.rho_bk_.dot(scaled_diff) < 0;
 }
 
-template <Update U, Direction D, typename S, class Generator>
-Span<S> combine(Random<S, Generator> &rng, Span<S> &&span_old,
-                Span<S> &&span_new, const Vec<S> &inv_mass, bool &uturn_flag) {
+template <Update U, Direction D, CombineResult R, typename S, class Generator>
+std::conditional_t<R == CombineResult::ThrowAway, std::optional<Span<S>>,
+                   std::pair<Span<S>, bool>>
+combine(Random<S, Generator> &rng, Span<S> &&span_old, Span<S> &&span_new,
```

Comment:
This code is getting very complicated.  I get what it's doing here, but it really feels like there should be a more elegant approach by refactoring at a higher level or pulling the U-turn test outside of `combine` in cases where we will always combine it whether there's a U-turn or not.

Reply:
I missed this suggestion to pull out the uturn test. It turns out we were already doing a redundant uturn test in the build_span functions before we called this, and I think it leads to the most readable code overall.